Http audio recording

hatton opened the pull request:
Audio Recording is now a pure backend thing which does not have a connection to the embedded browser.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/958)
<!-- Reviewable:end -->

JohnThomson commented:
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 191 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1m3SN0lI_j81NRd4t:-KC1m3SN0lI_j81NRd4u:-1214237557) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L191)):*
You've added a getPreviousSpan...shouldn't this use it?

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 219 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1mOsXDFrQIzJeDFKf:-KC1mOsXDFrQIzJeDFKg:541434695) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L219)):*
Seems weird to use get for a request that doesn't get anything (except I suppose a status) but does order the server to do a state-changing thing. Is there a better choice?

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 306 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1nq-yKeOcDUWOnSIM:-KC1nq-yKeOcDUWOnSIN:21720653) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L306)):*
Remove please.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 332 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1oFPzsE3E5f_IoxzJ:-KC1oFPzsE3E5f_IoxzK:-1280936503) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L332)):*
remove

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 381 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1oRJ-Lq6T49YGVfYG:-KC1oRJ-Lq6T49YGVfYH:-462099051) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L381)):*
Again 'get' seems weird.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 416 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1ojfjrKNby9u_LLD9:-KC1ojfke_2lbKouG-kZ:-1812606685) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L416)):*
remove at least 2 comment lines

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 427 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1p2RAWcrAvcs10N6k:-KC1p2RAWcrAvcs10N6l:-1446272995) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L427)):*
Comment needs updating...invoked because we set up a request to be notified of peak levels via a web socket...etc.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 874 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1pZKYdIRAlNUYIkaq:-KC1pZKYdIRAlNUYIkar:911223257) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L874)):*
I don't know exactly what toastr does, but not finding a recording is so normal I wonder whether we need to display a message at all.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 911 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1rmZFVkbCMDc4bQ83:-KC1rmZFVkbCMDc4bQ84:-1257983785) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L911)):*
This puzzled me for a long time (even though I probably wrote it originally) because I missed the + -label and it seemed redundant. A comment indicating that the 'expected' class is also needed on the label might help.

---

*[src/BloomBrowserUI/lib/toastr.css, line 1 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1X1kiy3YuZs8ZeSdg:-KC1X1kjP1Gy-Pyf4iHY:-353832147) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/lib/toastr.css#L1)):*
Did you mean to check in a css as well as its less??

---

*[src/BloomExe/Edit/AudioRecording.cs, line 114 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1tivdlroEymUyI4sY:-KC1tivdlroEymUyI4sZ:-283505927) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L114)):*
I don't know exactly why (senility?) but I read these calls as initializing the _peakLevelServer. Maybe a blank line would help? Or consider putting all the calls to the server argument before the rest of the init? It's such a common pattern to make an object and then initialize it, and both names have 'server'...

---

*[src/BloomExe/Edit/AudioRecording.cs, line 129 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1YT2Wddqa84KvDNxa:-KC1YT2XMCwRoToXoZAN:1240531120) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L129)):*
But this is saying, if the BOOK (audio folder) has any audio at all, you can try to listen to the audio for this page; but the PAGE may not have any at all.
There's existing code (in the epub maker, I think) that knows how to extract the audio spans from a DOM and test whether any of the corresponding audio files exist.

---

*[src/BloomExe/Edit/AudioRecording.cs, line 273 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1ZLFmdTcDs_t_a5KE:-KC1ZLFmdTcDs_t_a5KF:749328835) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L273)):*
I don't see the connection with this comment and the following code. It probably explains something important...did it get misplaced?? (It's duplicated below, where it seems to apply.)

---

*[src/BloomExe/Edit/AudioRecording.cs, line 298 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1Zij9bxTT9MDvbsM4:-KC1Zij9bxTT9MDvbsM5:-13119465) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L298)):*
Seems like everything else that can go wrong and we need to inform the user changed to Fails. Should this too? Maybe it's unlikely enough for it to be OK to keep the old handling...

---

*[src/BloomExe/Edit/AudioRecording.cs, line 485 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1_lI0yJsVNlYPZttq:-KC1_lI0yJsVNlYPZttr:1247540784) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L485)):*
Worth a mention here that it quietly succeeds if the file already doesn't exist?

---

*[src/BloomExe/Edit/AudioRecording.cs, line 499 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1_yPoLWBq8FMl6zOP:-KC1_yPoLWBq8FMl6zOQ:129668708) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L499)):*
Should we try a few times or something? I vaguely recall a SafeDelete method or similar...what happens if you try to delete a file that is currently being played?

---

*[src/BloomExe/Edit/EditingModel.cs, line 147 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1aYv_y4zUhsD5-JOG:-KC1aYv_y4zUhsD5-JOH:1689154950) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/EditingModel.cs#L147)):*
Is this redundant after setting _audioRecording.GetBookFolderPath to do the same? Or could that use this? Doesn't seem to be used.

---

*[src/BloomExe/web/BloomWebSocketServer.cs, line 10 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1avTv9uGycYHrYSSX:-KC1avTv9uGycYHrYSSY:2131651764) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/BloomWebSocketServer.cs#L10)):*
In my thinking, classes require comments explaining generally what they are responsible for and how they relate to other classes. In particular if this doesn't replace EnhancedImageServer you should explain the relationship. Consider that since I don't know what a web socket is, there's a good chance other readers won't either, so a summary explanation and/or link would be helpful. There's not much to this class, but reading through it almost everything is puzzling: who creates them? Is it a singleton? When is it disposed? Who opens and closes sockets? Who sends messages to them and why?

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 50 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1dtTXfnONxdVWI2rt:-KC1dtTXfnONxdVWI2ru:-489057623) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/EnhancedImageServer.cs#L50)):*
This seems to provide a second way of doing much the same thing. It may well be a better way...if so, could the couple of things that use the (non-simple?) request handlers be changed to use this?

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 216 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1dbw03z-uAKTFSFHB:-KC1dbw03z-uAKTFSFHC:371543040) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/EnhancedImageServer.cs#L216)):*
Since you moved these (before calling base), presumably order is important. If you still remember, how about capturing the reason it matters in a comment?

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 220 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1eOdH2wxzFXwOxc9Y:-KC1eOdH2wxzFXwOxc9Z:-1989879308) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/EnhancedImageServer.cs#L220)):*
It's interesting that you pass pair.Value (the key which localPath starts with) rather than the whole local path. That will prevent using this for anything where the path contains further useful information. Is that what makes this 'simple'? Otherwise it's VERY like what we do with _additionalRequestHandlers.

---

*[src/BloomExe/web/RequestInfo.cs, line 77 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1f1sQbg2abcpaeZoK:-KC1f1sQbg2abcpaeZoL:938297681) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/RequestInfo.cs#L77)):*
If this logging info was useful, a quick kludge would be to report except for .wav files, or except when 'audio' is in the path.

---

*[src/BloomExe/web/RequestInfo.cs, line 107 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1fjX6yAMRIpYkUyP7:-KC1fjX6yAMRIpYkUyP8:1224014842) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/RequestInfo.cs#L107)):*
How on earth did it compile before without this? Presumably there must have been some lastModified variable somewhere, not initialized to the right thing? Anyway, I don't know exactly what uses this, but should we worry that the file might not exist? (e.g., an audio one that has been cleared?)

---

*[src/BloomExe/web/SimpleHandler.cs, line 16 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1gIbOiIy6IS0J2vqj:-KC1gIbOiIy6IS0J2vqk:102062434) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L16)):*
Class and file name should generally match. If there's a reason to name the file after the delegate it should be commented.

---

*[src/BloomExe/web/SimpleHandler.cs, line 18 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1gjQnsQMBwbGsZbzc:-KC1gjQnsQMBwbGsZbzd:-342009849) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L18)):*
just remove if not needed?

---

*[src/BloomExe/web/SimpleHandler.cs, line 46 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1gzhZDWWmaE9Xgyay:-KC1gzhZDWWmaE9Xgyaz:513283803) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L46)):*
Not sure how often these are used, but seems it might clutter output window...maybe comment out for normal operation? (Also below)

---

*[src/BloomExe/web/SimpleHandler.cs, line 62 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1j1GHHE2lrgrEhLEN:-KC1j1GHHE2lrgrEhLEO:-1277473039) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L62)):*
typo (missing 'r' in Synchronizing).
Why is this better than Form.ActiveForm? Does the reason apply to any of the places I've used Form.ActiveForm?
I suppose this can't happen when the application has no open forms?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958)*
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: 11 of 20 files reviewed at latest revision, 27 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 219 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1mOsXDFrQIzJeDFKf:-KCH72x8_i4qrttDVNnj:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L219)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 306 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1nq-yKeOcDUWOnSIM:-KCH7HiC58rS_wHzDJl8:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L306)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 332 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1oFPzsE3E5f_IoxzJ:-KCH7Leaj_YEGlneE7B5:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L332)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 381 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1oRJ-Lq6T49YGVfYG:-KCH7RfUYgsi2u-CDyeu:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L381)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 416 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1ojfjrKNby9u_LLD9:-KCH7YknfNtXHBb5a5Jf:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L416)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 427 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1p2RAWcrAvcs10N6k:-KCH7kzmQXV-uG3vYpWS:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L427)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 874 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1pZKYdIRAlNUYIkaq:-KCH7psoRPNqlMu2moT8:-178352495) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L874)):*
You don't get an error just because it's not found. There would have to be an exception thrown on the c# side, while looking.

---

*[src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts, line 911 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1rmZFVkbCMDc4bQ83:-KCH8P7N-QfPYT9C3enB:739865769) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/audioRecording.ts#L911)):*
Not clear that when one of us missing what is there, that the appropriate response is to add a comment :-)

---

*[src/BloomExe/Edit/AudioRecording.cs, line 129 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1YT2Wddqa84KvDNxa:-KCHurVk9RBdQsekNT40:-1031311074) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L129)):*
Duh. Now just always says "yes".

---

*[src/BloomExe/Edit/AudioRecording.cs, line 273 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1ZLFmdTcDs_t_a5KE:-KCHv7ElB76M7vp2XeAu:-1054325829) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L273)):*
removed

---

*[src/BloomExe/Edit/AudioRecording.cs, line 298 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1Zij9bxTT9MDvbsM4:-KCHv_1aGyCqPdDWijGR:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L298)):*
Done.

---

*[src/BloomExe/Edit/AudioRecording.cs, line 485 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1_lI0yJsVNlYPZttq:-KCHvfbliPa7LEBq5vFT:1747121378) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L485)):*
I think the code shows that clearly. Not a public function.

---

*[src/BloomExe/Edit/AudioRecording.cs, line 499 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1_yPoLWBq8FMl6zOP:-KCHw5R9_GIx9qSHZazf:-1102417229) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/AudioRecording.cs#L499)):*
Would be a good enhancement. I don't see a SafeDelete around.

---

*[src/BloomExe/Edit/EditingModel.cs, line 147 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1aYv_y4zUhsD5-JOG:-KCHwP14v16j24_JOFg8:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/Edit/EditingModel.cs#L147)):*
Done.

---

*[src/BloomExe/web/BloomWebSocketServer.cs, line 10 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1avTv9uGycYHrYSSX:-KCHwV1ODbQBQVh81kmq:885766636) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/BloomWebSocketServer.cs#L10)):*
> Consider that since I don't know what a web socket is, there's a good chance other readers won't either,

Here's my view: all of us working in Bloom must regularly invest some time learning about web tech so that we get ideas that will help improve Bloom. Now, we can't know it all, so when we run into something we aren't aware of, then we google and expand our horizons.  WebSockets shipped in Firefox 4.0, so not a new thing.

But there is a lot to be said about how this use fits into our context, beyond reinventing mdn.org. I've added that comment and made the port into a parameter, which I hope makes it a bit clearer that this is just a low-level class. It isn't making claims about how it is used.

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 50 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1dtTXfnONxdVWI2rt:-KCI2YOc8h6tTk3QY9Zf:677266559) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/EnhancedImageServer.cs#L50)):*
Yup.

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 216 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1dbw03z-uAKTFSFHB:-KCI2xJWxdt55UhH8n9I:-1982088528) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/EnhancedImageServer.cs#L216)):*
This is just the more specific request; essentially these are how I'd like to see all api requests handled, getting them out of the Server *rat's nest*. So since these are specific, unambiguous requests, it makes sense to test for them first. Will add a comment.

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 220 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1eOdH2wxzFXwOxc9Y:-KCM6Sm5_Ykf0dN_vMJp:-2058478303) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/EnhancedImageServer.cs#L220)):*
Yup, I expect those will go away eventually. Following every thread to it's logical conclusion (which opens more threads) means I would never submit the PR :-)

---

*[src/BloomExe/web/RequestInfo.cs, line 107 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1fjX6yAMRIpYkUyP7:-KCLgAUkF3fzaPiOTejE:-704939160) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/RequestInfo.cs#L107)):*
From this and your last comment, I wonder if some differ was malfunctioning for you? This PR does not touch this part of the code.

---

*[src/BloomExe/web/SimpleHandler.cs, line 16 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1gIbOiIy6IS0J2vqj:-KCLgVISUyknAxSGqci4:160362544) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L16)):*
:-(

---

*[src/BloomExe/web/SimpleHandler.cs, line 18 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1gjQnsQMBwbGsZbzc:-KCLgqgWI4wj_FQ7gwdb:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L18)):*
Done.

---

*[src/BloomExe/web/SimpleHandler.cs, line 46 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1gzhZDWWmaE9Xgyay:-KCLh25KAmL-zVC5gDVo:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L46)):*
Done.

---

*[src/BloomExe/web/SimpleHandler.cs, line 62 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1j1GHHE2lrgrEhLEN:-KCLhRT4a_bXR8O3uo5u:633851719) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomExe/web/SimpleHandler.cs#L62)):*
http://stackoverflow.com/a/1127323/723299

---

*[src/BloomBrowserUI/lib/toastr.css, line 1 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958#-KC1X1kiy3YuZs8ZeSdg:-KCLigQDhYICks_pxwEM:-23700988) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/8aa442bd69d01dbe76cb11ecf7b5c3afcbdd1513/src/BloomBrowserUI/lib/toastr.css#L1)):*
I don't want either, since they are in node_modules. Removed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/958)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
One minor comment you might like to address while resolving comments...and there's still a typo in snchonize somewhere...but basically looks good.
